Cycle detected error instead of invalid function signature on trait

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: [Link to playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e3b9befe19a57230e9fc9f93c005bcf4)

```rust
fn main() {}

trait A {
    fn a(aa: B) -> Result<_, B> {
        Ok(())
    }
}

enum B {}
```

The current output is:

```
   Compiling s1 v0.1.0 (/Users/forest/Documents/test/s1)
error[E0391]: cycle detected when computing function signature of `A::a`
 --> src/main.rs:3:5
  |
3 |     fn a(aa: B) -> Result<_, B> {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: ...which requires type-checking `A::a`...
 --> src/main.rs:3:5
  |
3 |     fn a(aa: B) -> Result<_, B> {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing the variances of `B`...
 --> src/main.rs:8:1
  |
8 | enum B {}
  | ^^^^^^
  = note: ...which requires computing the variances for items in this crate...
  = note: ...which again requires computing function signature of `A::a`, completing the cycle
note: cycle used when collecting item types in top-level module
 --> src/main.rs:1:1
  |
1 | fn main() {}
  | ^^^^^^^^^

For more information about this error, try `rustc --explain E0391`.
error: could not compile `s1` due to previous error
```

I expected the compiler to let me know that function signature was wrong with the `_` in `Result<_, B>`. Once I changed this, it compiled fine.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"yanchen4791"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

The output of the test case after the fix is correct:
```
error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
 --> main.rs:3:27
  |
3 |     fn a(aa: B) -> Result<_, B> {
  |                    -------^----
  |                    |      |
  |                    |      not allowed in type signatures
  |                    help: replace with the correct return type: `Result<(), B>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0121`.
```

--- a/compiler/rustc_typeck/src/variance/mod.rs
+++ b/compiler/rustc_typeck/src/variance/mod.rs
@@ -37,6 +37,11 @@ fn crate_variances(tcx: TyCtxt<'_>, (): ()) -> CrateVariancesMap<'_> {
 }
 
 fn variances_of(tcx: TyCtxt<'_>, item_def_id: DefId) -> &[ty::Variance] {
+    // Skip items with no generics - there's nothing to infer in them.
+    if tcx.generics_of(item_def_id).count() == 0 {
+        return &[];
+    }
+
     match tcx.def_kind(item_def_id) {
         DefKind::Fn
         | DefKind::AssocFn

--- a/src/test/ui/typeck/issue-98260.rs
+++ b/src/test/ui/typeck/issue-98260.rs
@@ -1,0 +1,9 @@
+fn main() {}
+trait A {
+    fn a(aa: B) -> Result<_, B> {
+    //~^ ERROR: the placeholder `_` is not allowed within types on item signatures for return types [E0121]
+        Ok(())
+    }
+}
+
+enum B {}

--- a/src/test/ui/typeck/issue-98260.stderr
+++ b/src/test/ui/typeck/issue-98260.stderr
@@ -1,0 +1,12 @@
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for return types
+  --> $DIR/issue-98260.rs:3:27
+   |
+LL |     fn a(aa: B) -> Result<_, B> {
+   |                    -------^----
+   |                    |      |
+   |                    |      not allowed in type signatures
+   |                    help: replace with the correct return type: `Result<(), B>`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0121`.

